--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -84,15 +_,15 @@
 
     @Override
     public boolean test(final ItemStack input) {
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        if (predicate != null) {
+            return predicate.test(input.asBukkitCopy());
+        }
         // CraftBukkit start
         if (this.isExact()) {
             return this.itemStacks.contains(input); // Paper - Improve exact choice recipe ingredients (hashing FTW!)
         }
         // CraftBukkit end
-        // Purpur start - Add predicate to recipe's ExactChoice ingredient
-        if (predicate != null) {
-            return predicate.test(input.asBukkitCopy());
-        }
         // Purpur end - Add predicate to recipe's ExactChoice ingredient
         return input.is(this.values);
     }
@@ -104,7 +_,7 @@
             case io.papermc.paper.inventory.recipe.ItemOrExact.Item(final Holder<Item> item) ->
                 !this.isExact() && this.values.contains(item);
             case io.papermc.paper.inventory.recipe.ItemOrExact.Exact(final ItemStack exact) ->
-                this.isExact() && this.itemStacks.contains(exact);
+                    predicate != null ? predicate.test(exact.asBukkitCopy()) : this.isExact() && this.itemStacks.contains(exact); // Purpur - Add predicate to recipe's ExactChoice ingredient
         };
         // Paper end - Improve exact choice recipe ingredients
     }
