--- a/net/minecraft/world/level/portal/PortalShape.java
+++ b/net/minecraft/world/level/portal/PortalShape.java
@@ -30,6 +_,31 @@
     private static final int MIN_HEIGHT = 3;
     public static final int MAX_HEIGHT = 21;
     public static final BlockBehaviour.StatePredicate FRAME = (state, level, pos) -> state.is(Blocks.OBSIDIAN) || (org.purpurmc.purpur.PurpurConfig.cryingObsidianValidForPortalFrame && state.is(Blocks.CRYING_OBSIDIAN)); // Purpur - Crying obsidian valid for portal frames
+    // Tentacles start - Custom portals
+    // Map of blocks to use when creating destination return portals for a world
+    public static final java.util.Map<net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level>, Block> frameBlocks = java.util.Map.of(
+            ServerLevel.NETHER, Blocks.OBSIDIAN,
+            net.minecraft.resources.ResourceKey.create(
+                    net.minecraft.core.registries.Registries.DIMENSION,
+                    net.minecraft.resources.Identifier.withDefaultNamespace("amplified")), Blocks.BASALT,
+            net.minecraft.resources.ResourceKey.create(
+                    net.minecraft.core.registries.Registries.DIMENSION,
+                    net.minecraft.resources.Identifier.withDefaultNamespace("void")), Blocks.BLACKSTONE
+    );
+
+    // Map of predicates for valid portal frame blocks for a world
+    public static final java.util.Map<net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level>, BlockBehaviour.StatePredicate> framePredicates = java.util.Map.of(
+            ServerLevel.NETHER, FRAME,
+            net.minecraft.resources.ResourceKey.create(
+                    net.minecraft.core.registries.Registries.DIMENSION,
+                    net.minecraft.resources.Identifier.withDefaultNamespace("amplified")),
+            (state, level, pos) -> state.is(Blocks.BASALT),
+            net.minecraft.resources.ResourceKey.create(
+                    net.minecraft.core.registries.Registries.DIMENSION,
+                    net.minecraft.resources.Identifier.withDefaultNamespace("void")),
+            (state, level, pos) -> state.is(Blocks.BLACKSTONE)
+    );
+    // Tentacles end - Custom portals
     private static final float SAFE_TRAVEL_MAX_ENTITY_XY = 4.0F;
     private static final double SAFE_TRAVEL_MAX_VERTICAL_DELTA = 1.0;
     private final Direction.Axis axis;
@@ -55,40 +_,62 @@
     }
 
     public static Optional<PortalShape> findEmptyPortalShape(final LevelAccessor level, final BlockPos pos, final Direction.Axis preferredAxis) {
-        return findPortalShape(level, pos, shape -> shape.isValid() && shape.numPortalBlocks == 0, preferredAxis);
+        // Tentacles start - Custom portals - Allow specifying frame predicate
+        return findEmptyPortalShape(level, pos, preferredAxis, FRAME);
+    }
+
+    public static Optional<PortalShape> findEmptyPortalShape(final LevelAccessor level, final BlockPos pos, final Direction.Axis preferredAxis, final BlockBehaviour.StatePredicate framePredicate) {
+        return findPortalShape(level, pos, shape -> shape.isValid() && shape.numPortalBlocks == 0, preferredAxis, framePredicate);
+        // Tentacles end - Custom portals - Allow specifying frame predicate
     }
 
     public static Optional<PortalShape> findPortalShape(
         final LevelAccessor level, final BlockPos pos, final Predicate<PortalShape> isValid, final Direction.Axis preferredAxis
     ) {
         Optional<PortalShape> firstAxis = Optional.of(findAnyShape(level, pos, preferredAxis)).filter(isValid);
+        // Tentacles start - Custom portals - Allow specifying frame predicate
+        return findPortalShape(level, pos, isValid, preferredAxis, FRAME);
+    }
+
+    public static Optional<PortalShape> findPortalShape(
+            final LevelAccessor level, final BlockPos pos, final Predicate<PortalShape> isValid, final Direction.Axis preferredAxis, final BlockBehaviour.StatePredicate framePredicate
+    ) {
+        Optional<PortalShape> firstAxis = Optional.of(findAnyShape(level, pos, preferredAxis, framePredicate)).filter(isValid);
+        // Tentacles end - Custom portals - Allow specifying frame predicate
+
         if (firstAxis.isPresent()) {
             return firstAxis;
         } else {
             Direction.Axis otherAxis = preferredAxis == Direction.Axis.X ? Direction.Axis.Z : Direction.Axis.X;
-            return Optional.of(findAnyShape(level, pos, otherAxis)).filter(isValid);
+            return Optional.of(findAnyShape(level, pos, otherAxis, framePredicate)).filter(isValid); // Tentacles - Custom portals - Allow specifying frame predicate
         }
     }
 
     public static PortalShape findAnyShape(final BlockGetter level, final BlockPos pos, final Direction.Axis axis) {
+        // Tentacles start - Custom portals - Allow specifying frame predicate
+        return findAnyShape(level, pos, axis, FRAME);
+    }
+
+    public static PortalShape findAnyShape(final BlockGetter level, final BlockPos pos, final Direction.Axis axis, final BlockBehaviour.StatePredicate framePredicate) {
+        // Tentacles end - Custom portals - Allow specifying frame predicate
         org.bukkit.craftbukkit.util.BlockStateListPopulator blocks = new org.bukkit.craftbukkit.util.BlockStateListPopulator(((LevelAccessor) level).getMinecraftWorld()); // CraftBukkit
         Direction rightDir = axis == Direction.Axis.X ? Direction.WEST : Direction.SOUTH;
-        BlockPos bottomLeft = calculateBottomLeft(level, rightDir, pos, blocks); // CraftBukkit
+        BlockPos bottomLeft = calculateBottomLeft(level, rightDir, pos, blocks, framePredicate); // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
         if (bottomLeft == null) {
             return new PortalShape(axis, 0, rightDir, pos, 0, 0, blocks); // CraftBukkit
         } else {
-            int width = calculateWidth(level, bottomLeft, rightDir, blocks); // CraftBukkit
+            int width = calculateWidth(level, bottomLeft, rightDir, blocks, framePredicate); // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
             if (width == 0) {
                 return new PortalShape(axis, 0, rightDir, bottomLeft, 0, 0, blocks); // CraftBukkit
             } else {
                 MutableInt portalBlockCountOutput = new MutableInt();
-                int height = calculateHeight(level, bottomLeft, rightDir, width, portalBlockCountOutput, blocks); // CraftBukkit
+                int height = calculateHeight(level, bottomLeft, rightDir, width, portalBlockCountOutput, blocks, framePredicate); // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
                 return new PortalShape(axis, portalBlockCountOutput.intValue(), rightDir, bottomLeft, width, height, blocks); // CraftBukkit
             }
         }
     }
 
-    private static @Nullable BlockPos calculateBottomLeft(final BlockGetter level, final Direction rightDir, BlockPos pos, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks) { // CraftBukkit
+    private static @Nullable BlockPos calculateBottomLeft(BlockGetter level, Direction rightDir, BlockPos pos, org.bukkit.craftbukkit.util.BlockStateListPopulator blocks, final BlockBehaviour.StatePredicate framePredicate) { // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
         int minY = Math.max(level.getMinY(), pos.getY() - 21);
 
         while (pos.getY() > minY && isEmpty(level.getBlockState(pos.below()))) {
@@ -96,23 +_,23 @@
         }
 
         Direction leftDir = rightDir.getOpposite();
-        int edge = getDistanceUntilEdgeAboveFrame(level, pos, leftDir, blocks) - 1; // CraftBukkit
+        int edge = getDistanceUntilEdgeAboveFrame(level, pos, leftDir, blocks, framePredicate) - 1; // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
         return edge < 0 ? null : pos.relative(leftDir, edge);
     }
 
-    private static int calculateWidth(final BlockGetter level, final BlockPos bottomLeft, final Direction rightDir, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks) { // CraftBukkit
-        int width = getDistanceUntilEdgeAboveFrame(level, bottomLeft, rightDir, blocks); // CraftBukkit
+    private static int calculateWidth(final BlockGetter level, final BlockPos bottomLeft, final Direction rightDir, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks, final BlockBehaviour.StatePredicate framePredicate) { // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
+        int width = getDistanceUntilEdgeAboveFrame(level, bottomLeft, rightDir, blocks, framePredicate); // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
         return width >= 2 && width <= 21 ? width : 0;
     }
 
-    private static int getDistanceUntilEdgeAboveFrame(final BlockGetter level, final BlockPos pos, final Direction direction, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks) { // CraftBukkit
+    private static int getDistanceUntilEdgeAboveFrame(final BlockGetter level, final BlockPos pos, final Direction direction, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks, final BlockBehaviour.StatePredicate framePredicate) { // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
         BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
 
         for (int width = 0; width <= 21; width++) {
             blockPos.set(pos).move(direction, width);
             BlockState blockState = level.getBlockState(blockPos);
             if (!isEmpty(blockState)) {
-                if (FRAME.test(blockState, level, blockPos)) {
+                if (framePredicate.test(blockState, level, blockPos)) { // Tentacles - Custom portals - Allow specifying frame predicate
                     blocks.setBlock(blockPos, blockState, Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE); // CraftBukkit - lower left / right
                     return width;
                 }
@@ -120,7 +_,7 @@
             }
 
             BlockState belowState = level.getBlockState(blockPos.move(Direction.DOWN));
-            if (!FRAME.test(belowState, level, blockPos)) {
+            if (!framePredicate.test(belowState, level, blockPos)) { // Tentacles - Custom portals - Allow specifying frame predicate
                 break;
             }
             blocks.setBlock(blockPos, belowState, Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE); // CraftBukkit - bottom row
@@ -130,19 +_,20 @@
     }
 
     private static int calculateHeight(
-        final BlockGetter level, final BlockPos bottomLeft, final Direction rightDir, final int width, final MutableInt portalBlockCount, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks // CraftBukkit
+        final BlockGetter level, final BlockPos bottomLeft, final Direction rightDir, final int width, final MutableInt portalBlockCount, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks, final BlockBehaviour.StatePredicate framePredicate // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
     ) {
+
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
-        int height = getDistanceUntilTop(level, bottomLeft, rightDir, pos, width, portalBlockCount, blocks); // CraftBukkit
-        return height >= 3 && height <= 21 && hasTopFrame(level, bottomLeft, rightDir, pos, width, height, blocks) ? height : 0; // CraftBukkit
+        int height = getDistanceUntilTop(level, bottomLeft, rightDir, pos, width, portalBlockCount, blocks, framePredicate); // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
+        return height >= 3 && height <= 21 && hasTopFrame(level, bottomLeft, rightDir, pos, width, height, blocks, framePredicate) ? height : 0; // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
     }
 
     private static boolean hasTopFrame(
-        final BlockGetter level, final BlockPos bottomLeft, final Direction rightDir, final BlockPos.MutableBlockPos pos, final int width, final int height, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks // CraftBukkit
+        final BlockGetter level, final BlockPos bottomLeft, final Direction rightDir, final BlockPos.MutableBlockPos pos, final int width, final int height, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks, final BlockBehaviour.StatePredicate framePredicate // CraftBukkit // Tentacles - Custom portals - Allow specifying frame predicate
     ) {
         for (int i = 0; i < width; i++) {
             BlockPos.MutableBlockPos framePos = pos.set(bottomLeft).move(Direction.UP, height).move(rightDir, i);
-            if (!FRAME.test(level.getBlockState(framePos), level, framePos)) {
+            if (!framePredicate.test(level.getBlockState(framePos), level, framePos)) { // Tentacles - Custom portals - Allow specifying frame predicate
                 return false;
             }
             blocks.setBlock(framePos, level.getBlockState(framePos), Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE); // CraftBukkit - upper row
@@ -158,15 +_,16 @@
         final BlockPos.MutableBlockPos pos,
         final int width,
         final MutableInt portalBlockCount, final org.bukkit.craftbukkit.util.BlockStateListPopulator blocks // CraftBukkit
+        , final BlockBehaviour.StatePredicate framePredicate // Tentacles - Custom portals - Allow specifying frame predicate
     ) {
         for (int height = 0; height < 21; height++) {
             pos.set(bottomLeft).move(Direction.UP, height).move(rightDir, -1);
-            if (!FRAME.test(level.getBlockState(pos), level, pos)) {
+            if (!framePredicate.test(level.getBlockState(pos), level, pos)) { // Tentacles - Custom portals - Allow specifying frame predicate
                 return height;
             }
 
             pos.set(bottomLeft).move(Direction.UP, height).move(rightDir, width);
-            if (!FRAME.test(level.getBlockState(pos), level, pos)) {
+            if (!framePredicate.test(level.getBlockState(pos), level, pos)) { // Tentacles - Custom portals - Allow specifying frame predicate
                 return height;
             }
 
