--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -47,7 +_,7 @@
     protected final RegionBitmap usedSectors = new RegionBitmap();
 
     public RegionFile(final RegionStorageInfo info, final Path path, final Path externalFileDir, final boolean sync) throws IOException {
-        this(info, path, externalFileDir, RegionFileVersion.getCompressionFormat(), sync); // Paper - Configurable region compression format
+        this(info, path, externalFileDir, RegionFileVersion.getCompressionFormat(info.level()), sync); // Paper - Configurable region compression format // Tentacles - Per-world region compression format
     }
 
     public RegionFile(final RegionStorageInfo info, final Path path, final Path externalFileDir, final RegionFileVersion version, final boolean sync) throws IOException {
