--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -420,7 +_,7 @@
     // Paper end - rewrite chunk system
 
     public RegionFile(final RegionStorageInfo info, final Path path, final Path externalFileDir, final boolean sync) throws IOException {
-        this(info, path, externalFileDir, RegionFileVersion.getSelected(), sync);
+        this(info, path, externalFileDir, RegionFileVersion.getSelected(info.level()), sync); // Tentacles - Per-world region compression format
     }
 
     public RegionFile(final RegionStorageInfo info, final Path path, final Path externalFileDir, final RegionFileVersion version, final boolean sync) throws IOException {
