--- a/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
@@ -91,6 +_,16 @@
         return selected;
     }
 
+    // Tentacles start - Per-world region compression format
+    public static RegionFileVersion getSelected(String level) {
+        // Can't use PurpurWorldConfig here as WorldUpgrader (--forceUpgrade) happens before the world is loaded
+        //FIXME: getString might be slow?
+        String format = org.purpurmc.purpur.PurpurConfig.config
+                .getString("world-settings." + level + ".settings.compression-format", "default").toLowerCase();
+        return VERSIONS_BY_NAME.getOrDefault(format, getSelected());
+    }
+    // Tentacles end- Per-world region compression format
+
     public static boolean isValidVersion(final int version) {
         return VERSIONS.containsKey(version);
     }
