--- a/io/papermc/paper/inventory/recipe/StackedContentsExtrasMap.java
+++ b/io/papermc/paper/inventory/recipe/StackedContentsExtrasMap.java
@@ -11,11 +_,16 @@
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.item.crafting.Recipe;
 
+import java.util.HashSet;
+import java.util.function.Predicate;
+
 public final class StackedContentsExtrasMap {
 
     private final StackedContents<ItemOrExact> contents;
     public Object2IntMap<ItemOrExact.Item> regularRemoved = new Object2IntOpenHashMap<>(); // needed for re-using the regular contents (for ShapelessRecipe)
     public final ObjectSet<ItemStack> exactIngredients = new ObjectOpenCustomHashSet<>(ItemStackLinkedSet.TYPE_AND_TAG);
+    public Object2IntMap<ItemOrExact.Exact> predicateMatches = new Object2IntOpenHashMap<>(); // Purpur - Add predicate to recipe's ExactChoice ingredient
+    public final HashSet<Predicate<org.bukkit.inventory.ItemStack>> predicates = new HashSet<>(); // Purpur - Add predicate to recipe's ExactChoice ingredient
 
     public StackedContentsExtrasMap(final StackedContents<ItemOrExact> contents) {
         this.contents = contents;
@@ -23,10 +_,16 @@
 
     public void initialize(final Recipe<?> recipe) {
         this.exactIngredients.clear();
+        this.predicates.clear(); // Purpur - Add predicate to recipe's ExactChoice ingredient
         for (final Ingredient ingredient : recipe.placementInfo().ingredients()) {
             if (ingredient.isExact()) {
                 this.exactIngredients.addAll(ingredient.itemStacks());
             }
+            // Purpur start - Add predicate to recipe's ExactChoice ingredient
+            if (ingredient.predicate != null) {
+                this.predicates.add(ingredient.predicate);
+            }
+            // Purpur end - Add predicate to recipe's ExactChoice ingredient
         }
     }
 
@@ -51,14 +_,29 @@
         for (final ItemStack extra : this.exactIngredients) {
             this.contents.amounts.removeInt(new ItemOrExact.Exact(extra));
         }
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        for (final Object2IntMap.Entry<ItemOrExact.Exact> entry : this.predicateMatches.object2IntEntrySet()) {
+            this.contents.amounts.removeInt(entry);
+        }
+        // Purpur end - Add predicate to recipe's ExactChoice ingredient
         for (final Object2IntMap.Entry<ItemOrExact.Item> entry : this.regularRemoved.object2IntEntrySet()) {
             this.contents.amounts.addTo(entry.getKey(), entry.getIntValue());
         }
         this.exactIngredients.clear();
         this.regularRemoved.clear();
+        this.predicates.clear(); // Purpur - Add predicate to recipe's ExactChoice ingredient
     }
 
     public boolean accountStack(final ItemStack stack, final int count) {
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        if (this.predicates.stream()
+                .anyMatch(itemStackPredicate -> itemStackPredicate.test(stack.asBukkitCopy()))) {
+            final ItemOrExact.Exact exact = new ItemOrExact.Exact(stack);
+            predicateMatches.put(exact, count);
+            this.contents.account(exact, count);
+            return true;
+        }
+        // Purpur end - Add predicate to recipe's ExactChoice ingredient
         if (this.exactIngredients.contains(stack)) {
             this.contents.account(new ItemOrExact.Exact(stack), count);
             return true;
