--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -165,11 +_,16 @@
     protected void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston, net.minecraft.world.item.context.UseOnContext context) { // CraftBukkit - context
         if (!oldState.is(state.getBlock())) {
             if (inPortalDimension(level)) {
-                Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(level, pos, Direction.Axis.X);
-                if (optional.isPresent()) {
-                    optional.get().createPortalBlocks(level, (context == null) ? null : context.getPlayer()); // CraftBukkit - player
-                    return;
+                // Tentacles start - Custom portals - Check all frame predicates
+                for (StatePredicate predicate: PortalShape.framePredicates.values()) {
+                    Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(level, pos, Direction.Axis.X, predicate);
+
+                    if (optional.isPresent()) {
+                        optional.get().createPortalBlocks(level, (context == null) ? null : context.getPlayer()); // CraftBukkit - player
+                        return;
+                    }
                 }
+                // Tentacles end - Custom portals - Check all frame predicates
             }
 
             if (!state.canSurvive(level, pos)) {
@@ -205,23 +_,23 @@
             return false;
         } else {
             BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
-            boolean flag = false;
 
+            // Tentacles start - Custom portals  - Check all frame predicates
             for (Direction direction1 : Direction.values()) {
-                if (level.getBlockState(mutableBlockPos.set(pos).move(direction1)).is(Blocks.OBSIDIAN)) {
-                    flag = true;
-                    break;
+                BlockState state = level.getBlockState(mutableBlockPos.set(pos).move(direction1));
+                for (StatePredicate predicate: PortalShape.framePredicates.values()) {
+                    if (predicate.test(state, level, pos)) {
+                        Direction.Axis axis = direction.getAxis().isHorizontal()
+                        ? direction.getCounterClockWise().getAxis()
+                        : Direction.Plane.HORIZONTAL.getRandomAxis(level.random);
+
+                        return PortalShape.findEmptyPortalShape(level, pos, axis, predicate).isPresent();
+                    }
                 }
             }
 
-            if (!flag) {
-                return false;
-            } else {
-                Direction.Axis axis = direction.getAxis().isHorizontal()
-                    ? direction.getCounterClockWise().getAxis()
-                    : Direction.Plane.HORIZONTAL.getRandomAxis(level.random);
-                return PortalShape.findEmptyPortalShape(level, pos, axis).isPresent();
-            }
+            return false;
+            // Tentacles end - Custom portals - Check all frame predicates
         }
     }
 
