--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -107,7 +_,7 @@
         Direction.Axis updateAxis = directionToNeighbour.getAxis();
         Direction.Axis axis = state.getValue(AXIS);
         boolean wrongAxis = axis != updateAxis && updateAxis.isHorizontal();
-        return !wrongAxis && !neighbourState.is(this) && !PortalShape.findAnyShape(level, pos, axis).isComplete()
+        return !wrongAxis && !neighbourState.is(this) && PortalShape.framePredicates.values().stream().noneMatch(p -> PortalShape.findAnyShape(level, pos, axis).isComplete()) // Tentacles - Custom portals - check all frame predicates
             ? Blocks.AIR.defaultBlockState()
             : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
     }
@@ -145,7 +_,35 @@
 
     @Override
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
+        // Tentacles start - Custom portals - Check frame material, determine destination and return portal material
         ResourceKey<Level> newDimension = currentLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? Level.OVERWORLD : Level.NETHER; // CraftBukkit
+        BlockPos checkPos = portalEntryPos;
+        BlockState checkState;
+        Block frameBlock = null;
+        StatePredicate framePredicate = null;
+
+        do {
+            checkPos = checkPos.below();
+            checkState = currentLevel.getBlockState(checkPos);
+        } while(checkState.is(Blocks.NETHER_PORTAL) && checkPos.getY() > currentLevel.getMinY());
+
+        for (Map.Entry<ResourceKey<Level>, StatePredicate> entry : PortalShape.framePredicates.entrySet()) {
+            if (entry.getValue().test(checkState, currentLevel, checkPos)) {
+                // Destination world is the world associated with the frame material
+                // If we are already in that world then the destination is the Overworld
+                newDimension = currentLevel.dimension().equals(entry.getKey()) ? Level.OVERWORLD : entry.getKey();
+
+                // Ensure we find or create a matching return portal
+                // The destination portal should be made out of the material required to return to the original world
+                // E.g Basalt Amplified portal in the Nether should look for an Obsidian portal at the Amplified destination
+                framePredicate = PortalShape.framePredicates.getOrDefault(currentLevel.dimension(), PortalShape.framePredicates.getOrDefault(newDimension, PortalShape.FRAME));
+                frameBlock = PortalShape.frameBlocks.getOrDefault(currentLevel.dimension(), PortalShape.frameBlocks.getOrDefault(newDimension, Blocks.OBSIDIAN));
+                org.bukkit.Bukkit.getLogger().info("Destination world " + newDimension + " destination portal frameBlock " + frameBlock);
+                break;
+            }
+        }
+        // Tentacles end - Custom portals - Check frame material, determine destination and return portal material
+
         ServerLevel newLevel = currentLevel.getServer().getLevel(newDimension);
         // Paper start - Add EntityPortalReadyEvent
         io.papermc.paper.event.entity.EntityPortalReadyEvent portalReadyEvent = new io.papermc.paper.event.entity.EntityPortalReadyEvent(entity.getBukkitEntity(), newLevel == null ? null : newLevel.getWorld(), org.bukkit.PortalType.NETHER);
@@ -177,7 +_,7 @@
         newLevel = ((org.bukkit.craftbukkit.CraftWorld) result.to().getWorld()).getHandle();
         newWorldBorder = newLevel.getWorldBorder();
         approximateExitPos = newWorldBorder.clampToBounds(result.to().getX(), result.to().getY(), result.to().getZ());
-        return this.getExitPortal(newLevel, entity, portalEntryPos, approximateExitPos, toNether, newWorldBorder, result);
+        return this.getExitPortal(newLevel, entity, portalEntryPos, approximateExitPos, toNether, newWorldBorder, result, framePredicate, frameBlock); // Tentacles - Custom portals - Allow specifying frame predicate and material
         // CraftBukkit end
     }
 
@@ -189,8 +_,10 @@
         final boolean toNether,
         final WorldBorder worldBorder
         , org.bukkit.craftbukkit.event.PortalEventResult result // CraftBukkit
+        , @Nullable StatePredicate framePredicate // Tentacles - Custom portal frame
+        , @Nullable Block frameBlock // Tentacles - Custom portal frame
     ) {
-        Optional<BlockPos> exitPortalPos = newLevel.getPortalForcer().findClosestPortalPosition(approximateExitPos, worldBorder, result.searchRadius()); // CraftBukkit
+        Optional<BlockPos> exitPortalPos = newLevel.getPortalForcer().findClosestPortalPosition(approximateExitPos, worldBorder, result.searchRadius(), framePredicate); // CraftBukkit // Tentacles - Custom portals - Allow limiting to specific frame material
         BlockUtil.FoundRectangle exitPortal;
         TeleportTransition.PostTeleportTransition post;
         if (exitPortalPos.isPresent()) {
@@ -207,7 +_,7 @@
             post = TeleportTransition.PLAY_PORTAL_SOUND.then(e -> e.placePortalTicket(pos));
         } else if (result.canCreatePortal()) { // CraftBukkit
             Direction.Axis sourcePortalAxis = entity.level().getBlockState(portalEntryPos).getOptionalValue(AXIS).orElse(Direction.Axis.X);
-            Optional<BlockUtil.FoundRectangle> createdExit = newLevel.getPortalForcer().createPortal(approximateExitPos, sourcePortalAxis, entity, result.createRadius()); // CraftBukkit
+            Optional<BlockUtil.FoundRectangle> createdExit = newLevel.getPortalForcer().createPortal(approximateExitPos, sourcePortalAxis, entity, result.createRadius(), frameBlock); // CraftBukkit // Tentacles - Custom portals - Allow specifying frame material
             if (createdExit.isEmpty()) {
                 // LOGGER.error("Unable to create a portal, likely target out of worldborder"); // CraftBukkit
                 return null;
